### ISSUE TITLE
Sharing context/state with multiple screens (multiple roots, wix/react-native-navigation)

### ISSUE THREAD

**Issue by mrousavy** [facebookexperimental/Recoil#719]
As far as I understood, there has to be exactly one `<RecoilRoot>` throughout the entire app, because only state under the same `RecoilRoot` tree is shared. I've tried to use Recoil for my React Native app, but since I'm using [wix/react-native-navigation](https://github.com/wix/react-native-navigation), which uses separate roots for each Screen, I couldn't manage to get Recoil working correctly, since state is only persisted per-screen basis.

I don't have a lot of experience with state management libraries, but other libraries such as Redux, MobX, ... also somehow manage to work with react-native-navigation, that's why I'm creating this issue. I might also be doing something wrong, if so, please let me know.

**Comment by mrousavy** [comment]
Currently, I am registering each screen with a custom Recoil HoC I've written. So e.g.:

```ts
RegisterScreen('HomeScreen', () => recoilHoC(HomeScreen));
RegisterScreen('ProfileScreen', () => recoilHoC(ProfileScreen));
```

should both share the same context/recoil state.

My `recoilHoC` looks like this:

```ts
import React, { ComponentType } from 'react';
import hoistNonReactStatics from 'hoist-non-react-statics';
import { RecoilRoot } from 'recoil';

export default function recoilHoC<TProps>(Component: ComponentType): ComponentType {
  function Wrapper(props: TProps): React.ReactElement {
    return (
      <RecoilRoot>
        <Component {...props} />
      </RecoilRoot>
    );
  }

  Wrapper.displayName = `recoilHoC(${Component.displayName || Component.name})`;

  hoistNonReactStatics(Wrapper, Component);

  return Wrapper as ComponentType;
}
``` 

I've tried to play around with [`useRecoilBridgeAcrossReactRoots`](https://recoiljs.org/docs/api-reference/core/useRecoilBridgeAcrossReactRoots), and came up with this HoC:

```ts
import React, { ComponentType } from 'react';
import hoistNonReactStatics from 'hoist-non-react-statics';
import { RecoilRoot, useRecoilBridgeAcrossReactRoots_UNSTABLE } from 'recoil';

export default function recoilHoC<TProps>(Component: ComponentType): ComponentType {
  function Bridge(props: TProps): React.ReactElement {
    const RecoilBridge = useRecoilBridgeAcrossReactRoots_UNSTABLE();
    return (
      <RecoilBridge>
        <Component {...props} />
      </RecoilBridge>
    );
  }
  Bridge.displayName = `useRecoilBridgeAcrossReactRoots_UNSTABLE(${Component.displayName || Component.name})`;
  hoistNonReactStatics(Bridge, Component);

  function Wrapper(props: TProps): React.ReactElement {
    return (
      <RecoilRoot>
        <Bridge {...props} />
      </RecoilRoot>
    );
  }
  Wrapper.displayName = `recoilHoC(${Component.displayName || Component.name})`;
  hoistNonReactStatics(Wrapper, Component);

  return Wrapper as ComponentType;
}
```

but obviously I'm missing something because I still have separate `<RecoilRoot>`s for each screen.

**Comment by drarmstr** [comment]
The point of [`useRecoilBridgeAcrossReactRoots_UNSTABLE()`](https://recoiljs.org/docs/api-reference/core/useRecoilBridgeAcrossReactRoots) is to support shared synchronized state with different React trees without them having shared React Context.  

Please note from the example in the documentation that the hook must be used from within a `<RecoilRoot>` (`<MyApp />` calls `<Bridge>`) and it provides a bridge component that can be used elsewhere that will have shared state with that host root.  In your code above the `Wrapper` is attempting to call `useRecoilBridgeAcrossReactRoots_UNSTABLE()` outside of a `<RecoilRoot>`, hence the error.

**Comment by mrousavy** [comment]
@drarmstr Thanks for your response, awesome name btw. Okay so if I were to use `<RecoilRoot>` in my HomeScreen, I can then use the generated `RecoilBridge` for every other screen? The problem with that is, that this requires my HomeScreen to be rendered, and will crash if it unmounts (e.g. on logout -> root change).

**Comment by drarmstr** [comment]
I'm really not very familiar with React Native or `wix/react-native-navigation`..  Would it be possible to put `<RecoilRoot>` at a higher level and use the bridge in the home screen as well as the other screens?

**Comment by mrousavy** [comment]
@drarmstr No, since there is no "higher level". It basically creates a new react app for each screen.

**Comment by mrousavy** [comment]
@drarmstr since this is a fundamental design problem (React.Context API not being share-able across _screens_), do you want to close this issue or are you positive we will find a solution to this?

**Comment by riordanpawley** [comment]
@mrousavy Did you get anywhere on this issue?
I'm just stumbling into it now as I'm investigating whether RNN is compatible with recoil.
From this line in RNN's docs ["If you need to trigger a re-render across all screens, there are many popular third party libraries such as MobX or Redux"](https://wix.github.io/react-native-navigation/docs/third-party-react-context): I think it might be possible to get recoil working but I'm not sure.
Recoil might be too intertwined into the react component tree to do so :(

**Comment by mrousavy** [comment]
@riordanpawley I ended up writing my own small state management library based on callbacks and hooks, which works fine for me using RNN. 
Unfortunately Recoil didn't work because it uses React Context, and RNN has a separate react component tree per screen (so also separate react context) - if Recoil was using something else like proxies it would work

**Comment by a-eid** [comment]
I'm wondering if there is any plans to support / work around this ?

**Comment by mondaychen** [comment]
@a-eid we are discussing about a possible solution to support multiple similar use cases.

**Comment by BenjaBobs** [comment]
Couldn't you observe every change to basically "clone" the state into an object that lives outside the React context, and then initialize every `<RecoilRoot>` with that object? This is of course assuming the render trees of each screen is new each time you navigate, which I don't know if is the case.

**Comment by mrousavy** [comment]
I don't really get what the benefits of using the React.Context API are, mind elaborating here? @mondaychen @drarmstr 

Maybe we can figure something out if we know what features we're looking for. I honestly don't understand why one would need to use React.Context API instead of just using Callbacks to trigger the state updates.

**Comment by doublelam** [comment]
Hi, anything new about this issue?

**Comment by Xooone47** [comment]
anything updated?

**Comment by guyca** [comment]
This issue is caused by a conscious design design, so it seems unlikely it will ever get fixed. If someone is limited by this aspect of Recoil, I'd suggest migrating to a different library. Something like [Mobx](https://github.com/mobxjs/mobx), [Jotai](https://github.com/pmndrs/jotai), or [Obsidian](https://github.com/wix-incubator/obsidian) will do the job.

**Comment by Xooone47** [comment]
> 

btw, zustand supports this feature
